python/pyspark/pandas/frame.py

```diff
@@ -8164,6 +8165,137 @@ def update(self, other: "DataFrame", join: str = "left", overwrite: bool = True)
+        cats -1.000000  1.666667
+
+        >>> np.random.seed(42)
+        >>> df = ps.DataFrame(np.random.randn(1000, 5),
+        ...                   columns=['a', 'b', 'c', 'd', 'e'])
+        >>> df.cov()
+                  a         b         c         d         e
+        a  0.998438 -0.020161  0.059277 -0.008943  0.014144
+        b -0.020161  1.059352 -0.008543 -0.024738  0.009826
+        c  0.059277 -0.008543  1.010670 -0.001486 -0.000271
+        d -0.008943 -0.024738 -0.001486  0.921297 -0.013692
+        e  0.014144  0.009826 -0.000271 -0.013692  0.977795
+
+        **Minimum number of periods**
+
+        This method also supports an optional ``min_periods`` keyword
+        that specifies the required minimum number of non-NA observations for
+        each column pair in order to have a valid result:
+
+        >>> np.random.seed(42)
+        >>> df = pd.DataFrame(np.random.randn(20, 3),
+        ...                   columns=['a', 'b', 'c'])
+        >>> df.loc[df.index[:5], 'a'] = np.nan
+        >>> df.loc[df.index[5:10], 'b'] = np.nan
+        >>> sdf = ps.from_pandas(df)
+        >>> sdf.cov(min_periods=12)
+                  a         b         c
+        a  0.316741       NaN -0.150812
+        b       NaN  1.248003  0.191417
+        c -0.150812  0.191417  0.895202
+        """
+        min_periods = 1 if min_periods is None else min_periods
+
+        psdf = self[
+            [
+                col
+                for col in self.columns
+                if isinstance(self[col].spark.data_type, BooleanType)
+                or (
+                    isinstance(self[col].spark.data_type, NumericType)
+                    and not isinstance(self[col].spark.data_type, DecimalType)
+                )
+            ]
+        ]
+
+        num_cols = len(psdf.columns)
+        data_cols = psdf._internal.data_spark_column_names
+
+        def _loop(func: Callable) -> None:
+            step = 0
+            for r in range(0, num_cols):
+                step += r
+                for c in range(r, num_cols):
+                    func(r, c, step)
+
+        count_not_null_scols = []
+        _loop(
+            lambda r, c, _: count_not_null_scols.append(
+                F.count(
+                    F.when(F.col(data_cols[r]).isNotNull() & F.col(data_cols[c]).isNotNull(), 1)
+                )
+            ),
+        )
+
+        count_not_null = (
+            psdf._internal.spark_frame.replace(float("nan"), None)
+            .select(*count_not_null_scols)
+            .head(1)[0]
+        )
+
+        cov_scols = []
+        _loop(
+            lambda r, c, step: cov_scols.append(
+                F.covar_samp(F.col(data_cols[r]).cast("double"), F.col(data_cols[c]).cast("double"))
+                if count_not_null[r * num_cols + c - step] >= min_periods
+                else F.lit(None)
+            ),
+        )
+        pair_cov = psdf._internal.spark_frame.select(*cov_scols).head(1)[0]
```

Comment:
Can we do this from here to below without using a Spark job?

@dgd-contributor it would be great to show some intermediate and explain it with some comments.  It's a bit difficult to follow the logic.

Reply:
> Can we do this from here to below without using a Spark job?

Yes, from here to below, we only work with list having length ```(no_colums^2) / 2```

> @dgd-contributor it would be great to show some intermediate and explain it with some comments. It's a bit difficult to follow the logic.

I write some comments, could you take a look? Thank you.